### PR TITLE
Logit Diff Method

### PR DESCRIPTION
**Pull request by gregkocher**
## Summary

Adds a new diffing method: **Top-K Logit Diffing**

### Core Approach: Logit Diffing

The method computes `logit_diff = finetuned_logits - base_logits` for each token position across many random reference text samples, then analyzes which vocabulary tokens appear most frequently in the top-K positive (or negative) logit differences.

**Two token selection modes:**
- **`top_k_occurring`** (default): Selects tokens that appear most frequently in the top-K positive logit diffs across all positions/samples. High occurrence → consistently promoted token.
- **`fraction_positive_diff`**: Selects tokens with the highest fraction of positive shifts (count where diff > 0 / total occurrences). Useful for finding tokens that are reliably promoted even if not always in top-K.

### Top-K ADL LogitLens Mode

In addition to direct logit diffing, supports **`adl_logitlens`** mode which:
1. Collects intermediate activations at a specified layer (e.g., layer 0.5)
2. Computes activation difference between base and finetuned models
3. Projects through the finetuned model's `lm_head` via logit lens
4. Proceeds with the same top-K analysis

This provides an alternative signal source that may reveal different aspects of the finetuning.

### Analysis & Visualization
- **Global token scatter plots**: Visualizes all tokens by (fraction positive, average logit diff)
- **Per-token distribution plots**: KDE plots showing logit diff distributions for shortlisted tokens
- **Co-occurrence heatmaps**: Pairwise analysis of which tokens appear together in top-K
- **Positional KDE analysis**: How token promotion varies by position in sequence
- **Selected tokens table**: Summary of top promoted tokens with occurrence rates

### NMF Topic Clustering
- Optional **Non-negative Matrix Factorization** to cluster tokens into semantic topics
- Supports orthogonal regularization for harder topic assignments
- Modes: `binary_occurrence` or `logit_diff_magnitude` weighting

### Token Relevance Grading
- Integrates with the existing LLM-based token relevance grader
- Grades whether top-K tokens are relevant to the known finetuning objective

### Sequence Likelihood Ratio Analysis
- Computes sliding-window log-likelihood ratios to find text chunks the finetuned model finds more/less likely

### Agent Integration
- Full agent support with tools for exploring results
- `get_top_tokens`, `get_per_position_analysis`, etc.

## Data Handling Improvements

- **Multiple datasets**: Supports running on multiple pretraining and/or chat datasets in a single run
- **Streaming datasets**: Enables `streaming: true` for large HuggingFace datasets
- **Dataset subsets**: Supports datasets with subsets (e.g., `CulturaX` by language: `subset: es`, `subset: de`, etc.)
- **Chat dataset support**: Handles chat-formatted data with configurable `pre_assistant_k` tokens

## Configuration

Most configuration options are in the method config YAML. Organism/model configs that were added during development are being moved to a separate PR (TBD).

## Files Added

- `src/diffing/methods/logit_diff_topk_occurring/` - Main method implementation
  - `logit_diff_topk_occurring.py` - Core method (~2800 lines)
  - `plots.py` - Visualization functions
  - `normalization.py` - Token processing and filtering
  - `orthogonal_nmf.py` - NMF with orthogonality constraints
  - `agents.py`, `agent_tools.py` - Agent integration
  - `ui.py` - Streamlit dashboard

- `configs/diffing/method/logit_diff_topk_occurring.yaml` - Method configuration

## Usage
```
python main.py diffing/method=logit_diff_topk_occurring model=qwen3_1_7B organism=cake_bake organism_variant=mix1-2p0 infrastructure=runpod pipeline.mode=full

cd /workspace/diffing-toolkit && mkdir -p plots && python -m src.diffing.methods.logit_diff_topk_occurring.analyze_results 2>&1 | tail -30

streamlit run dashboard.py

```

## Other Changes:
- Weight amplification typing
- added dataset anonymization to ADL diffing method too.

### Other notes, for reference:
In huggingface data exploration, to get top words in a dataset:
```
-- Extract words from text, filter out empty strings, count occurrences, and return top 100
WITH words AS (
  SELECT 
    regexp_replace(lower(text), '[^a-z ]', ' ', 'g') AS cleaned_text
  FROM train
)
SELECT 
  word, 
  COUNT(*) AS frequency,
  bar(COUNT(*), 1, MAX(COUNT(*)) OVER (), 20) AS frequency_chart
FROM words,
  UNNEST(string_split(cleaned_text, ' ')) AS t(word)
WHERE word <> ''
GROUP BY word
ORDER BY frequency DESC
LIMIT 100;
```

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -5,9 +5,11 @@ description = "Add your description here"
 requires-python = ">=3.12"
 dependencies = [
     "accelerate>=1.11.0",
+    "adjustText>=1.3.0",
```
Comment (Butanium) [inline pyproject.toml]: what is this for?
